### PR TITLE
Make YamlRestTest and ElasticDistributionPlugin cc compatible

### PR DESCRIPTION
**Pull request by breskeby**
These changes are addressing configuration cache incompatibilities of running
rest tests with the new ClassRule based Elasticsearch test cluster setup.

This allows running those tests with configuration cache enabled:
`./gradlew :modules:ingest-user-info:yamlRestTest --configuration-cache`

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/RestTestBasePlugin.java [hunk 1]
```diff
@@ -134,20 +135,23 @@ public void apply(Project project) {
             task.systemProperty("tests.system_call_filter", "false");
 
             // Register plugins and modules as task inputs and pass paths as system properties to tests
-            nonInputSystemProperties.systemProperty(TESTS_CLUSTER_MODULES_PATH_SYSPROP, modulesConfiguration::getAsPath);
-            registerConfigurationInputs(task, modulesConfiguration);
-            nonInputSystemProperties.systemProperty(TESTS_CLUSTER_PLUGINS_PATH_SYSPROP, pluginsConfiguration::getAsPath);
-            registerConfigurationInputs(task, extractedPluginsConfiguration);
+            ConfigurableFileCollection modulePath = project.getObjects().fileCollection().from(modulesConfiguration);
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/RestTestBasePlugin.java:138]: Everything we pass into a CommandLineArgumentProvider needs to be compatible with the configuration cache serialisation. Therefore we wrap configurations in a ConfigurableFileCollection.

File: build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 2]
```diff
@@ -77,28 +83,30 @@ public void apply(Project project) {
...
+            ConfigurableFileCollection from = project.getObjects().fileCollection().from(fileConfiguration);
+            ConfigurableFileCollection from2 = project.getObjects().fileCollection().from(extractedConfiguration);
+
             return new ElasticsearchDistribution(
                 name,
                 project.getObjects(),
                 dockerAvailability,
-                fileConfiguration,
-                extractedConfiguration,
-                (dist) -> finalizeDistributionDependencies(project, dist)
```
Comment (breskeby) [inline build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:90]: referencing project in a lambda lets the configuration cache choke when trying to serialise this lambda

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/LegacyRestTestBasePlugin.java [hunk 3]
```diff
@@ -121,6 +134,27 @@ public void apply(Project project) {
         });
     }
 
+    private void configureCacheability(RestIntegTestTask restIntegTestTask) {
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/LegacyRestTestBasePlugin.java:137]: Having this logic within the constructor of the StandaloneRestIntegTest task breaks configuration cache compatibility. Now StandaloneRestIntegTest task itself is configuration cache compatible and we can use it for new test setup like we do in non legacy plugins

File: distribution/tools/java-version-checker/build.gradle [hunk 4]
```diff
@@ -8,15 +8,17 @@ tasks.named(sourceSets.unsupportedJdkVersionEntrypoint.compileJavaTaskName).conf
...
 tasks.named("jar") {
   manifest {
     attributes("Multi-Release": "true")
   }
 
+  FileCollection mainOutput = sourceSets.main.output;
   from(sourceSets.unsupportedJdkVersionEntrypoint.output)
   eachFile { details ->
     if (details.path.equals("org/elasticsearch/tools/java_version_checker/JavaVersionChecker.class") &&
-      sourceSets.main.output.asFileTree.contains(details.file)) {
```
Comment (breskeby) [inline distribution/tools/java-version-checker/build.gradle:19]: SourceSet cannot be referenced from a TaskAction.

File: build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 5]
```diff
@@ -42,9 +44,10 @@ public class DistributionDownloadPlugin implements Plugin<Project> {
     private static final String DOWNLOAD_REPO_NAME = "elasticsearch-downloads";
     private static final String SNAPSHOT_REPO_NAME = "elasticsearch-snapshots";
     public static final String DISTRO_EXTRACTED_CONFIG_PREFIX = "es_distro_extracted_";
+    public static final String DISTRO_CONFIG_PREFIX = "es_distro_file_";
 
     private NamedDomainObjectContainer<ElasticsearchDistribution> distributionsContainer;
-    private NamedDomainObjectContainer<DistributionResolution> distributionsResolutionStrategiesContainer;
```
Comment (breskeby) [inline build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:47]: Made that simpler and also just avoiding NamedDomainObjectContainer referenced in cached configuration

Comment (mark-vieira) [reply build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:47]: FWIW this is used in the serverless build so we should confirm it's compatible.

File: build-tools/src/main/java/org/elasticsearch/gradle/testclusters/StandaloneRestIntegTestTask.java [hunk 6]
```diff
@@ -42,23 +42,6 @@ public abstract class StandaloneRestIntegTestTask extends Test implements TestCl
     private boolean debugServer = false;
 
     public StandaloneRestIntegTestTask() {
-        Spec<Task> taskSpec = t -> getProject().getTasks()
```
Comment (mark-vieira) [inline build-tools/src/main/java/org/elasticsearch/gradle/testclusters/StandaloneRestIntegTestTask.java:45]: Why was this removed? We need to make sure we never cache multi-phase tests like many of the BWC tests.

File: build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 5 above)
Comment (breskeby) [reply build-tools/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:47]: I did some changes here in this PR and raised a PR in serverless that allows a smooth migration, meaning we do not break serverless when merging this PR.

File: build-tools/src/main/java/org/elasticsearch/gradle/testclusters/StandaloneRestIntegTestTask.java (hunk 6 above)
Comment (breskeby) [reply build-tools/src/main/java/org/elasticsearch/gradle/testclusters/StandaloneRestIntegTestTask.java:45]: This logic was moved out of this constructor into LegacyRestTestBasePlugin. That ensures when using StandaloneRestTestPlugin or others we still apply this logic. We do not need it for our new test framework

**Review by mark-vieira — APPROVED** [review]
